File: src/editor/tag_dictionary_fs.rs
use godot::{
    engine::{DirAccess, ResourceLoader},
    prelude::*,
};

use crate::tag_dictionary::TagDictionary;

#[derive(GodotClass)]
#[class(tool, init, base = RefCounted)]
pub struct TagDictionaryFs {
    base: Base<RefCounted>,
    dictionaries: Array<Gd<TagDictionary>>,
}

#[godot_api]
impl TagDictionaryFs {
    fn _read_dir_recursive(&self, dir: &str) -> PackedStringArray {
        let mut output = PackedStringArray::new();

        if let Some(mut dir_access) = DirAccess::open(dir.into_godot()) {
            dir_access.list_dir_begin();

            let mut filename = dir_access.get_next();

            while !filename.is_empty() {
                if dir_access.current_is_dir() {
                    if !filename.to_string().starts_with(".") {
                        let sub_dir = format!("{}/{}", dir, filename);
                        let sub_dir_files = self._read_dir_recursive(&sub_dir);

                        for file in sub_dir_files.as_slice() {
                            output.push(file.to_godot().clone());
                        }
                    }
                } else {
                    output.push(GString::from(format!("{}/{}", dir, filename)));
                }

                filename = dir_access.get_next();
            }
        }

        output
    }

    /// Returns the stored dictionaries.
    #[func]
    pub fn get_dictionaries(&self) -> Array<Gd<TagDictionary>> {
        self.dictionaries.clone()
    }

    #[func]
    pub fn scan_fs(&mut self, from_dir: GString) {
        let resources = self._read_dir_recursive(&from_dir.to_string());
        let mut loader = ResourceLoader::singleton();

        for resource_path in resources.as_slice() {
            if resource_path.to_string().ends_with(".res")
                || resource_path.to_string().ends_with(".tres")
            {
                if let Some(resource) = loader.load(resource_path.clone()) {
                    if let Ok(dictionary) = resource.try_cast::<TagDictionary>() {
                        self.dictionaries.push(dictionary);
                    }
                }
            }
        }
    }
}

File: src/editor/TagDictionaryFS.cpp
//
// Created by Lasagnaking on 19/03/2025.
//

#include "TagDictionaryFS.h"
#include <godot_cpp/classes/dir_access.hpp>
#include <godot_cpp/classes/resource_loader.hpp>

using namespace octod::gameplay::tags::editor;

static TagDictionaryFS *singleton = nullptr;

TagDictionaryFS * TagDictionaryFS::get_singleton()
{
	if (singleton == nullptr)
	{
		singleton = memnew(TagDictionaryFS);
	}

	return singleton;
}

TypedArray<octod::gameplay::tags::TagDictionary> TagDictionaryFS::get_dictionaries() const
{
	return dictionaries;
}

void TagDictionaryFS::scan_file_system(const String &p_from_directory)
{
	PackedStringArray resources = _read_directory_recursive(p_from_directory);
	ResourceLoader *resource_loader = ResourceLoader::get_singleton();

	for (int i = 0; i < resources.size(); i++)
	{
		if (const String& res_path = resources[i]; res_path.ends_with(".res") || res_path.ends_with(".tres"))
		{
			if (Ref<Resource> resource = resource_loader->load(res_path); resource.is_valid())
			{
				if (const TagDictionary *maybe_tag_dictionary = cast_to<TagDictionary>(resource.ptr()); maybe_tag_dictionary != nullptr)
				{
					dictionaries.push_back(maybe_tag_dictionary);
				}
			}
		}
	}
}

void TagDictionaryFS::_bind_methods()
{
}

PackedStringArray TagDictionaryFS::_read_directory_recursive(const String &p_directory) // NOLINT(*-no-recursion)
{
	PackedStringArray output;

	const Ref<DirAccess> dir_access = DirAccess::open(p_directory);

	ERR_FAIL_COND_V_MSG(dir_access->list_dir_begin() != OK, output, "Error while reading directory: " + p_directory);

	String filename = dir_access->get_next();

	while (!filename.is_empty())
	{
		if (dir_access->current_is_dir())
		{
			if (!filename.begins_with("."))
			{
				PackedStringArray sub_dir_files = _read_directory_recursive(vformat("{%s}/{%s}", p_directory, filename));

				for (int i = 0; i < sub_dir_files.size(); i++)
				{
					output.push_back(sub_dir_files[i]);
				}
			}
		} else
		{
			output.push_back(vformat("{%s}/{%s}", p_directory, filename));
		}

		filename = dir_access->get_next();
	}

	return output;
}
